### PR TITLE
FileMonitor redesign

### PR DESCRIPTION
**Pull request by JosJuice**
Advantages:

* Simpler code in general
* No extra volume objects created
* Now actually notices if the disc or partition gets changed while the core is running
* No longer picks up on disc access done by the GUI (it used to do so as long as the core was running)
* Gets rid of a Core dependency in DiscIO

There's one performance disadvantage:

* FileMonitor is now a bit slower when used with VolumeDirectory because FileMonitor now always uses the FileSystemGCWii code for finding filenames instead of VolumeDirectory finding the filename on its own and directly hooking into FileMonitor.  But this isn't such a big deal, because it's happening on the DVD thread, and my currently unmerged file system PR will make FileSystemGCWii's file finding code about as fast as VolumeDirectory's.

### CODE REVIEW

File: Source/Core/Core/FileMonitor.h [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+class IFileSystem;
+class IVolume;
+}
+
+namespace FileMonitor
+{
+// Can be called with nullptr to set the file system to nothing. When not called
+// with nullptr, the volume must remain valid until the next SetFileSystem call.
+void SetFileSystem(const DiscIO::IVolume* volume);
+// SetFileSystem must have been called before calling Log
```
Comment (iwubcode) [inline Source/Core/Core/FileMonitor.h]: Should we assert that s_filesystem is nullptr in this function?

Comment (JosJuice) [inline Source/Core/Core/FileMonitor.h]: No. I can't see why that would make sense.

Comment (iwubcode) [inline Source/Core/Core/FileMonitor.h]: Log says that "SetFileSystem" must have been called.  Ensuring that s_filesystem is nullptr would potentially be a way to do that?

**Comment by JMC47** [comment]
Question: Will this fix Filemonitor tanking performance in DTK audio titles?

File: Source/Core/Core/FileMonitor.h (hunk 1 above)
Comment (JosJuice) [inline Source/Core/Core/FileMonitor.h]: Oh, you mean asserting it in `Log`? We can't do that. The requirement is that `SetFileSystem` has been called at some point and that the passed-in volume is still valid (unless it was `nullptr`). When `Log` is called more than once after each `SetFileSystem` call (which happens all the time), `s_filesystem` will not be `nullptr` – this is completely normal behavior.

Actually, the current comment might be a little bit misleading. It's not that calling `Log` without `SetFileSystem` is forbidden, it's just that it won't be able to log anything, like if someone would have called `SetFileSystem` with `nullptr` manually. I'll change it.

**Comment by JosJuice** [comment]
@JMC47 No. That's addressed by the DTK queue PR (merged) and filesystem redesign PR (not merged).

File: Source/Core/Core/HW/DVDInterface.cpp [hunk 2]
```diff
@@ -576,7 +586,9 @@ void SetLidOpen(bool open)
 bool ChangePartition(u64 offset)
 {
   DVDThread::WaitUntilIdle();
-  return s_inserted_volume->ChangePartition(offset);
+  bool success = s_inserted_volume->ChangePartition(offset);
```
Comment (lioncash) [inline Source/Core/Core/HW/DVDInterface.cpp]: `const bool`

File: Source/Core/Core/FileMonitor.cpp [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+#include "Core/FileMonitor.h"
+
+#include <algorithm>
+#include <cctype>
+#include <memory>
+#include <string>
+#include <unordered_set>
+
+#include "Common/CommonTypes.h"
+#include "Common/Logging/LogManager.h"
```
Comment (lioncash) [inline Source/Core/Core/FileMonitor.cpp]: Missed this one the previous review, but `Common/Logging/Log.h` also needs to be included for the logging #defines. After that, it looks fine to merge.

Comment (JosJuice) [inline Source/Core/Core/FileMonitor.cpp]: Done.

**Comment by JosJuice** [comment]
@dolphin-emu-bot rebuild

**Pull request merged**